crates/marketplace/src/builder_state.rs

```diff
@@ -172,7 +155,206 @@ pub struct BuilderState<TYPES: NodeType> {
+        return existing_states;
+    }
+
+    // At this point, we don't have any "ideal" matches or scenarios.  So we
+    // need to look for a suitable fall-back. The best fallback condition to
+    // start with is any [BuilderState] that has the maximum spawned view
+    // number whose value is smaller than the current [QuorumProposal].
+    let maximum_stored_view_number_smaller_than_quorum_proposal = global_state_read_lock
+        .spawned_builder_states
+        .keys()
+        .map(|builder_state_id| *builder_state_id.parent_view)
+        .filter(|view_number| view_number < &current_view_number)
+        .max();
+
+    // If we have a maximum view number that meets our criteria, then we should
+    // return all [BuilderStateId]s that match this view number.
+    // This can lead to multiple [BuilderStateId]s being returned.
+    if let Some(maximum_stored_view_number_smaller_than_quorum_proposal) =
+        maximum_stored_view_number_smaller_than_quorum_proposal
+    {
+        // If we are the maximum stored view number smaller than the quorum
+        // proposal's view number, then we are the best fit.
+        let mut result = HashSet::new();
+        for builder_state_id in
+            global_state_read_lock
+                .spawned_builder_states
+                .keys()
+                .filter(|builder_state_id| {
+                    builder_state_id.parent_view.u64()
+                        == maximum_stored_view_number_smaller_than_quorum_proposal
+                })
+        {
+            result.insert(builder_state_id.clone());
+        }
+        return result;
+    }
+
+    // This is our last ditch effort to continue making progress.  If there is
+    // only one [BuilderState] active, then we should return that as the best
+    // fit, as it will be the only way we can continue making progress with
+    // the builder.
+    if global_state_read_lock.spawned_builder_states.len() == 1 {
+        let mut result = HashSet::new();
+        for builder_state_id in global_state_read_lock.spawned_builder_states.keys() {
+            result.insert(builder_state_id.clone());
+        }
+        return result;
+    }
+
+    // This implies that there are only larger [BuilderState]s active than
+    // the one we are.  This is weird, it implies that some sort of time
+    // travel has occurred view-wise.  It is unclear what to do in this
+    // situation.
+
+    HashSet::new()
+}
+
 impl<TYPES: NodeType> BuilderState<TYPES> {
+    /// [am_i_the_best_builder_state_to_extend] is a utility method that
+    /// attempts to determine whether we are among the best [BuilderState]s to
+    /// extend from.
+    async fn am_i_the_best_builder_state_to_extend(
+        &self,
+        quorum_proposal: Arc<Proposal<TYPES, QuorumProposal<TYPES>>>,
+    ) -> bool {
+        let best_builder_states_to_extend =
+            best_builder_states_to_extend(quorum_proposal.clone(), self.global_state.clone()).await;
```

Comment:
Makes sense. I think for this test we probably don't need all the info from `best_builder_states_to_extend`, but for future tests or debugging it may be very helpful. I added this issue for us to determine whether to simplify it after other refactor we want to do: https://github.com/EspressoSystems/marketplace-builder-core/issues/124. But for now, let's keep it!